docs/AQL/master03-syntax.adoc

```diff
@@ -451,7 +451,17 @@ AQL structure has the following clauses and these clauses must be listed in the
 
 === FROM
 
-The `FROM` clause utilises class expressions and a set of containment criteria to specify the data source from which the query required data is to be retrieved.  Its function is similar as the `FROM` clause of an SQL expression.
+The `FROM` clause is used to specify the domain of the query, that is a subset of the universe of data that could be retrieve from a CDR. That universe is anything defined in the openEHR RM (that is a set of classes that models clinical records), and archetypes (that define specific data sets based on the openEHR RM). Because of that, the `FROM` clause uses `class expressions` to specify the subsets of data, giving context to the query.
```

Comment:
I'd suggest not changing the definition of from clause for now. We need to discuss and define the scope of FROM, because it is not all of RM, it is a subset of it, mostly due to most common && sensible usage patterns, language design and finally performance. The mention of RM and archetypes complicates things and not entirely accurate IMHO: we're using RM attributes in AQL, it is not like TDO/TDS/Flattened JSON, i.e. using archetype level paths.

Reply:
Yes it's a bit tricky. Firstly, AQL isn't / shouldn't be specific to any RM or kind of data - it works with any archetyped data in any kind of persistent store - not just a CDR. So if there are any specific parts of some RM that are not used in AQL queries, it is only because the AQL view of such an RM is just a piece of it (i.e. the AQL environment literally doesn't know about certain parts of an RM). So nothing special needs to be done to achieve this - it's just a question of what RM is made visible in an AQL environment (in future: should be via BMM).

I'd be inclined to minimise changes to the text for 1.0.1 - we can get more radical later. How about:

The `FROM` clause is used to specify the domain of the query, that is a subset of the universe of data that could be retrieved from a persistent store of archetyped data based on an Information Model (IM). That universe is anything defined in the IM that may have been constrained by archetypes known in the AQL environment. Accordingly, the `FROM` clause uses `class expressions` and containment expressions to specify the subsets of possibly matching data to provide the context of a query.

Reply:
If you suggest we keep the change, then I'd suggest we also add the following. 

The IM types used in the `class expressions` are currently a vendor specific subset of the IM. i.e. only a particular set of IM types are supported in the `FROM` clause and this set can differ between vendors' AQL implementations.

Reply:
That's undoubtedly true in reality, but it's not actually part of the spec if you think about it; i.e. it doesn't constitute information that a vendor can use to know what to implement.

But we can restate that in a form something like: the sub-set of any published information model that is used as the basis for AQL queries over the data created according to that model is a matter for the model user community to agree on. From the point of view of AQL, the model visible to it is the entire model.

However, I think this probably only makes sense to state in 1.1.0, because right now, we have no formal statement about how an IM is even made visible to AQL, or what subset of the openEHR RM this specifically is. This is all technically achievable with BMM, but we have not yet connected BMM with AQL unfortunately.

Reply:
Well, you seem to have arrived my original suggestion: let's not comment on what FROM exposes from IM because it is complicated. To clarify: I'm not keen to say anything about this at the moment, but if we'll get into what FROM exposes, let's at least mention that it is 'in reality' not the whole IM. I'm happy with your rewording but I'm in favour of not getting into this detail in the first place

Reply:
On reflection, I think you are probably right. @ppazos can we pull this change. Your explanation is better in some bits, but you can see from this discussion we need to get rid of the openEHR- and RM-specificity  - but current AQL just doesn't have anything to say about that so I am inclined to agree with @serefarikan and not touch for now (even though it will drive your perfectionist sensibilities off a cliff - mine too ;)

Reply:
@wolandscat @serefarikan the idea of the change to the FROM definition was really to improve the previous one, not changing the meaning. Did the same for the SELECT. Also think we need to formally define the rest, like the WHERE and other clauses. For instance there are pending changes from our discussion about TIMEWINDOW.

We can pull this out for now.

The mention of the RM is because we are using the RM classes in the FROM (EHR, COMPOSITION, ENTRIEs, and so on). But if you are thinking about being able to use AQL for non-openEHR RMs, which is something I didn't consider, we should add that to the spec, since it is not mentioned anywhere, and I think that is misleading by absence.

IMO we need to provide more detail, and current definition is too open, so open that doesn't add any value. The lack of definition on the previous version of the spec really causes many implementation issues, leading to inconsistent/not compatible implementations. Adding how each clause should be interpreted, giving examples and ideas of implementation is better for implementers., which I think it's something we need to provide.

> The IM types used in the class expressions are currently a vendor specific subset of the IM. i.e. only a particular set of IM types are supported in the FROM clause and this set can differ between vendors' AQL implementations.

I'm OK with that, just two things:

1. to mention "class expression" that should be defined before it's use.
2. we need to say which is the minimal set or the required set of IM classes supported in the FROM, and also say if AQL support different RMs, because today different vendors might support different RM classes in AQL, and the standard shouldn't IMO leave everything open for any implementation. One of the selling point of AQL is portability, which in reality is far from happening, and part is because the issues in the current spec, vendors just tried their best and added extensions for missing things, which we need to harmonize on a new and more complete spec.

Definitely we need a SEC call to talk about these points.

Reply:
@ppazos  I agree with your points re the problems caused by the vagueness of current definition of FROM. We'll need to do quite a bit of work to fix this though. Since we all agree to hold this for the moment, I'm resolving this discussion.